internal/bucket/lifecycle/lifecycle.go

```diff
@@ -309,99 +310,175 @@ func (o ObjectOpts) ExpiredObjectDeleteMarker() bool {
+		case DeleteAction, DeleteVersionAction:
+			return true
+		}
+		switch es[j].EventAction {
+		case DeleteAction, DeleteVersionAction:
+			return false
+		}
+		return true
+	}
+
+	// Prefer earlier occurring event
+	return es[i].Due.Before(es[j].Due)
+}
+
+// eval returns the lifecycle event applicable at now. If now is the zero value of time.Time, it returns the upcoming lifecycle event.
+func (lc Lifecycle) eval(obj ObjectOpts, now time.Time) lifecycleEvent {
+	var events []lifecycleEvent
 	if obj.ModTime.IsZero() {
-		return action
+		return lifecycleEvent{}
 	}
+
+	// Handle expiry of restored object; NB Restored Objects have expiry set on
+	// them as part of RestoreObject API. They aren't governed by lifecycle
+	// rules.
+	if !obj.RestoreExpires.IsZero() && now.After(obj.RestoreExpires) {
+		action := DeleteRestoredAction
+		if !obj.IsLatest {
+			action = DeleteRestoredVersionAction
+		}
+
+		events = append(events, lifecycleEvent{
+			EventAction: action,
+			Due:         now,
+		})
+	}
+
 	for _, rule := range lc.FilterActionableRules(obj) {
 		if obj.ExpiredObjectDeleteMarker() {
 			if rule.Expiration.DeleteMarker.val {
 				// Indicates whether MinIO will remove a delete marker with no noncurrent versions.
 				// Only latest marker is removed. If set to true, the delete marker will be expired;
 				// if set to false the policy takes no action. This cannot be specified with Days or
 				// Date in a Lifecycle Expiration Policy.
-				return DeleteVersionAction
+				events = append(events, lifecycleEvent{
+					EventAction: DeleteVersionAction,
+					RuleID:      rule.ID,
+					Due:         now,
+				})
+				// No other conflicting actions apply to an expired object delete marker
+				break
 			}
 
 			if !rule.Expiration.IsDaysNull() {
 				// Specifying the Days tag will automatically perform ExpiredObjectDeleteMarker cleanup
 				// once delete markers are old enough to satisfy the age criteria.
 				// https://docs.aws.amazon.com/AmazonS3/latest/userguide/lifecycle-configuration-examples.html
-				if time.Now().UTC().After(ExpectedExpiryTime(obj.ModTime, int(rule.Expiration.Days))) {
-					return DeleteVersionAction
+				if expectedExpiry := ExpectedExpiryTime(obj.ModTime, int(rule.Expiration.Days)); now.After(expectedExpiry) {
+					events = append(events, lifecycleEvent{
+						EventAction: DeleteVersionAction,
+						RuleID:      rule.ID,
+						Due:         expectedExpiry,
+					})
+					// No other conflicting actions apply to an expired object delete marker
+					break
 				}
 			}
 		}
 
-		if !rule.NoncurrentVersionExpiration.IsDaysNull() {
-			// Skip rules with newer noncurrent versions specified.
-			// These rules are not handled at an individual version
-			// level. ComputeAction applies only to a specific
-			// version.
-			if !obj.IsLatest && rule.NoncurrentVersionExpiration.NewerNoncurrentVersions > 0 {
-				continue
-			}
-			if obj.VersionID != "" && !obj.IsLatest && !obj.SuccessorModTime.IsZero() {
```

Comment:
It seems this works with a non versioned bucket, but neither obj.VersionID != "" or !obj.IsLatest  are well defined in case of a non versioned bucket. I think it is better to have a separate option to indicate this is a versioned bucket

Reply:
@vadmeste I ran a local test by creating an object in an unversioned bucket using MinIO release `RELEASE.2019-12-19T22-52-26Z` (i.e legacy object) with the following ILM policy.
```
{
 "Rules": [
  {
   "ID": "after 1 days",
   "NoncurrentVersionExpiration": {
    "NoncurrentDays": 1,
    "NewerNoncurrentVersions": 0
   },
   "Status": "Enabled"
  }
 ]
}
```
This test was to check if `NoncurrentVersionExpiration` rules incorrectly applied to an object with no noncurrent version. The test passed, i.e the object had `IsLatest` set to `true` making it ineligible for noncurrent version specific actions.